p2p/protocol/holepunch/svc.go

```diff
@@ -257,5 +266,7 @@ func (s *Service) handleNewStream(str network.Stream) {
 // TODO: find a solution for this.
 func (s *Service) DirectConnect(p peer.ID) error {
 	<-s.hasPublicAddrsChan
+	s.holePuncherMx.Lock()
+	defer s.holePuncherMx.Unlock()
```

Comment:
This locks the mutex for a **very** long time (on the order of 10s or so). Save `s.holePuncher` to a local variable instead.

Reply:
Or maybe just use RMutex? And call RLock?